File: src/parsers.py
import io
import os
import re
import pypdf
import docx


def clean_pdf_text(text):
    """
    Clean up PDF font ligature and kerning glitches.
    Preserves single-letter words like 'a', 'A', 'I' cleanly.
    """
    if not text:
        return ""

    # Remove null bytes and unprintable control characters
    text = text.replace('\x00', '')
    text = re.sub(r'[\x01-\x08\x0b\x0c\x0e-\x1f]', '', text)

    # Fix specific recurring PDF OCR / kerning glitches
    glitches = {
        "r ainw ater": "rainwater",
        "rainw ater": "rainwater",
        "collec tion": "collection",
        "s ystem": "system",
        "t ypically": "typically",
        "suppor ts": "supports",
        "sanitar y": "sanitary",
        "inspec tion": "inspection",
        "r un-of f": "run-off",
        "coef ficient": "coefficient",
        "over flow": "overflow",
        "ver min": "vermin",
        "ventil ation": "ventilation",
        "ac tivit y": "activity",
        "oper ation": "operation",
        "distr ic t": "district",
        "prov ince": "province",
        "v ill age": "village",
        "par ish": "parish",
        "descr ibe": "describe",
        "af fec ted": "affected",
        "r ainfall": "rainfall",
        "gut ter": "gutter",
        "gut ter ing": "guttering"
    }

    for glitch, fix in glitches.items():
        text = re.sub(re.escape(glitch), fix, text, flags=re.IGNORECASE)

    # Clean up footnote superscript markers like 'source.a' -> 'source.'
    text = re.sub(r'([a-z0-9])\.([a-d1-9])\b', r'\1.', text)

    # Clean excessive whitespace
    text = re.sub(r'[ \t]+', ' ', text)
    text = re.sub(r'\n\s*\n+', '\n\n', text)

    return text.strip()


def fallback_extract_pdf_text(file_bytes_or_path, file_name="document.pdf"):
    """Fallback text extraction for malformed or corrupted PDFs using raw stream parsing."""
    if isinstance(file_bytes_or_path, bytes):
        raw_bytes = file_bytes_or_path
    else:
        try:
            with open(file_bytes_or_path, "rb") as f:
                raw_bytes = f.read()
        except Exception:
            return []

    # Find text sequences inside parentheses or stream blocks
    matches = re.findall(rb'\(([^()\\\r\n]{3,})\)', raw_bytes)
    extracted_strings = []
    for match in matches:
        try:
            s = match.decode('utf-8', errors='ignore').strip()
            if len(s) > 3 and any(c.isalnum() for c in s):
                extracted_strings.append(s)
        except Exception:
            continue

    full_text = " ".join(extracted_strings)
    full_text = clean_pdf_text(full_text)

    if not full_text:
        return []

    # Split into chunks of ~1500 chars to emulate pages
    chunk_size = 1500
    documents = []
    paragraphs = full_text.split(" ")
    current_chunk = []
    current_length = 0
    page_num = 1

    for word in paragraphs:
        current_chunk.append(word)
        current_length += len(word) + 1
        if current_length >= chunk_size:
            chunk_text = " ".join(current_chunk).strip()
            if chunk_text:
                documents.append({
                    "text": chunk_text,
                    "metadata": {
                        "source": file_name,
                        "page": page_num
                    }
                })
                page_num += 1
            current_chunk = []
            current_length = 0

    if current_chunk:
        chunk_text = " ".join(current_chunk).strip()
        if chunk_text:
            documents.append({
                "text": chunk_text,
                "metadata": {
                    "source": file_name,
                    "page": page_num
                }
            })

    return documents


def parse_pdf(file_bytes_or_path, file_name="document.pdf"):
    """Extract text from PDF pages with clean text repair and non-strict error recovery."""
    documents = []

    try:
        if isinstance(file_bytes_or_path, bytes):
            reader = pypdf.PdfReader(io.BytesIO(file_bytes_or_path), strict=False)
        else:
            reader = pypdf.PdfReader(file_bytes_or_path, strict=False)

        try:
            num_pages = len(reader.pages)
        except Exception:
            num_pages = 0

        for i in range(num_pages):
            try:
                page = reader.pages[i]
                text = page.extract_text() or ""
            except Exception:
                text = ""

            text = clean_pdf_text(text)

            if text:
                documents.append({
                    "text": text,
                    "metadata": {
                        "source": file_name,
                        "page": i + 1
                    }
                })
    except Exception as e:
        print(f"[parse_pdf] pypdf failed to parse '{file_name}': {e}")

    # Fallback to stream extraction if pypdf yields no valid text or encounters fatal syntax errors
    if not documents:
        documents = fallback_extract_pdf_text(file_bytes_or_path, file_name=file_name)

    return documents


def parse_docx(file_bytes_or_path, file_name="document.docx"):
    """Extract text from Word DOCX document."""
    documents = []

    if isinstance(file_bytes_or_path, bytes):
        doc = docx.Document(io.BytesIO(file_bytes_or_path))
    else:
        doc = docx.Document(file_bytes_or_path)

    full_text = []
    for para in doc.paragraphs:
        if para.text.strip():
            full_text.append(para.text.strip())

    text = "\n".join(full_text)
    if text:
        documents.append({
            "text": text,
            "metadata": {
                "source": file_name,
                "page": 1
            }
        })

    return documents


def parse_txt(file_bytes_or_path, file_name="document.txt"):
    """Extract text from plain text or Markdown document."""
    documents = []

    if isinstance(file_bytes_or_path, bytes):
        text = file_bytes_or_path.decode("utf-8", errors="ignore")
    else:
        with open(file_bytes_or_path, "r", encoding="utf-8", errors="ignore") as f:
            text = f.read()

    text = text.strip()
    if text:
        documents.append({
            "text": text,
            "metadata": {
                "source": file_name,
                "page": 1
            }
        })

    return documents


def parse_document(file_name, file_bytes_or_path):
    """Unified document parser for PDF, DOCX, TXT, and MD files."""
    ext = os.path.splitext(file_name)[1].lower()

    if ext == ".pdf":
        return parse_pdf(file_bytes_or_path, file_name=file_name)
    elif ext in [".docx", ".doc"]:
        return parse_docx(file_bytes_or_path, file_name=file_name)
    elif ext in [".txt", ".md", ".csv", ".json", ".log"]:
        return parse_txt(file_bytes_or_path, file_name=file_name)
    else:
        return parse_txt(file_bytes_or_path, file_name=file_name)

File: lib/parsers.ts
import mammoth from "mammoth";
import pdfParse from "pdf-parse";
import { DocumentChunk } from "./types";

/**
 * Clean up OCR, PDF kerning glitches, and formatting artifacts.
 */
export function cleanText(text: string): string {
  if (!text) return "";

  let cleaned = text
    .replace(/\0/g, "")
    .replace(/[\x01-\x08\x0b\x0c\x0e-\x1f]/g, "")
    .replace(/\/H\d+/g, "")
    .replace(/■\s*\d+/g, "")
    .replace(/P\.T\.O\s*\d*/gi, "");

  const glitches: Record<string, string> = {
    "r ainw ater": "rainwater",
    "rainw ater": "rainwater",
    "collec tion": "collection",
    "s ystem": "system",
    "t ypically": "typically",
    "suppor ts": "supports",
    "sanitar y": "sanitary",
    "inspec tion": "inspection",
    "r un-of f": "run-off",
    "coef ficient": "coefficient",
    "over flow": "overflow",
    "ver min": "vermin",
    "ventil ation": "ventilation",
    "ac tivit y": "activity",
    "oper ation": "operation",
    "distr ic t": "district",
    "prov ince": "province",
    "v ill age": "village",
    "descr ibe": "describe",
    "af fec ted": "affected",
    "r ainfall": "rainfall",
    "gut ter": "gutter",
    "gut ter ing": "guttering"
  };

  for (const [glitch, fix] of Object.entries(glitches)) {
    const reg = new RegExp(glitch, "gi");
    cleaned = cleaned.replace(reg, fix);
  }

  // Remove footnote superscript artifacts e.g. "term.a" -> "term."
  cleaned = cleaned.replace(/([a-z0-9])\.([a-d1-9])\b/g, "$1.");

  // Normalize excessive spacing
  cleaned = cleaned.replace(/[ \t]+/g, " ");
  cleaned = cleaned.replace(/\n\s*\n+/g, "\n\n");

  return cleaned.trim();
}

/**
 * Chunk text into overlapping segments preserving sentence & heading boundaries.
 */
export function chunkText(
  fullText: string,
  sourceName: string,
  chunkSize = 1000,
  overlap = 150,
  basePage = 1
): DocumentChunk[] {
  const cleaned = cleanText(fullText);
  if (!cleaned) return [];

  const chunks: DocumentChunk[] = [];
  const paragraphs = cleaned.split("\n\n");

  let currentBuffer: string[] = [];
  let currentLength = 0;
  let chunkIndex = basePage;
  let activeHeading = "";

  for (const para of paragraphs) {
    const trimmedPara = para.trim();
    if (!trimmedPara) continue;

    // Detect section or lecture headings
    const headingMatch = trimmedPara.match(/^(?:LECTURE\s*[-–—:]?\s*\d+|Module\s*[-–—:]?\s*\w+|[A-Z][A-Za-z\s]{3,35}:)/);
    if (headingMatch && trimmedPara.length < 80) {
      activeHeading = headingMatch[0].replace(/:$/, "").trim();
    }

    if (currentLength + trimmedPara.length + 2 > chunkSize && currentBuffer.length > 0) {
      let chunkString = currentBuffer.join("\n\n").trim();
      
      // If the chunk doesn't have the active heading at the start, prepend context breadcrumb
      if (activeHeading && !chunkString.toLowerCase().includes(activeHeading.toLowerCase())) {
        chunkString = `[Section: ${activeHeading}]\n${chunkString}`;
      }

      if (chunkString.length > 30) {
        chunks.push({
          id: `${sourceName}-chunk-${chunkIndex}`,
          source: sourceName,
          page: chunkIndex,
          text: chunkString
        });
        chunkIndex++;
      }

      // Compute overlap from end of current buffer
      const lastText = currentBuffer[currentBuffer.length - 1];
      if (lastText && lastText.length > overlap) {
        currentBuffer = [lastText.slice(-overlap), trimmedPara];
        currentLength = overlap + trimmedPara.length + 2;
      } else {
        currentBuffer = [trimmedPara];
        currentLength = trimmedPara.length;
      }
    } else {
      currentBuffer.push(trimmedPara);
      currentLength += trimmedPara.length + 2;
    }
  }

  if (currentBuffer.length > 0) {
    let chunkString = currentBuffer.join("\n\n").trim();
    if (activeHeading && !chunkString.toLowerCase().includes(activeHeading.toLowerCase())) {
      chunkString = `[Section: ${activeHeading}]\n${chunkString}`;
    }
    if (chunkString.length > 20) {
      chunks.push({
        id: `${sourceName}-chunk-${chunkIndex}`,
        source: sourceName,
        page: chunkIndex,
        text: chunkString
      });
    }
  }

  return chunks;
}

/**
 * Multi-format document parser.
 */
export async function parseDocument(
  fileName: string,
  buffer: Buffer
): Promise<{ text: string; chunks: DocumentChunk[] }> {
  const ext = fileName.slice(fileName.lastIndexOf(".")).toLowerCase();
  let extractedText = "";
  const chunks: DocumentChunk[] = [];

  if (ext === ".pdf") {
    try {
      const pageTexts: { page: number; text: string }[] = [];
      const pagerender = function (pageData: any) {
        return pageData.getTextContent().then((textContent: any) => {
          let text = "";
          let lastY = 0;
          for (const item of textContent.items) {
            if (lastY === item.transform[5] || !lastY) {
              text += item.str;
            } else {
              text += "\n" + item.str;
            }
            lastY = item.transform[5];
          }
          pageTexts.push({ page: pageData.pageIndex + 1, text: cleanText(text) });
          return text;
        });
      };

      const data = await pdfParse(buffer, { pagerender });
      extractedText = data.text || "";

      // If page-by-page texts were extracted, chunk per page
      if (pageTexts.length > 0) {
        for (const pt of pageTexts) {
          if (pt.text && pt.text.length > 20) {
            const pageChunks = chunkText(pt.text, fileName, 1200, 150, pt.page);
            chunks.push(...pageChunks);
          }
        }
      }
    } catch (err) {
      console.warn(`PDF parse error for ${fileName}:`, err);
      extractedText = buffer.toString("utf-8").replace(/[^\x20-\x7E\n\r\t]/g, " ");
    }
  } else if (ext === ".docx" || ext === ".doc") {
    try {
      const result = await mammoth.extractRawText({ buffer });
      extractedText = result.value || "";
    } catch (err) {
      console.warn(`DOCX parse error for ${fileName}:`, err);
      extractedText = buffer.toString("utf-8");
    }
  } else {
    // Plain text, Markdown, CSV, JSON, etc.
    extractedText = buffer.toString("utf-8");
  }

  // If page-by-page didn't produce chunks, chunk full text
  if (chunks.length === 0) {
    const defaultChunks = chunkText(extractedText, fileName);
    chunks.push(...defaultChunks);
  }

  return {
    text: extractedText,
    chunks
  };
}
